File: server/automation_service/src/main/java/com/chatterly/automation_service/entity/Automation.java
package com.chatterly.automation_service.entity;

import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonManagedReference;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import lombok.Data;

@Data
@Entity
@Table(name = "automation")
public class Automation implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Column(nullable = false, columnDefinition = "varchar(255) default 'Untitled'")
    private String name;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    @Column(nullable = false, columnDefinition = "boolean default false")
    private boolean active;

    @OneToMany(mappedBy = "automation", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    @JsonManagedReference
    private List<Trigger> triggers;

    @OneToOne(mappedBy = "automation", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    @JsonManagedReference
    private Listener listener;

    @OneToMany(mappedBy = "automation", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    @JsonManagedReference
    private List<Post> posts;

    @OneToMany(mappedBy = "automation", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    @JsonManagedReference
    private List<Dms> dms;

    @OneToMany(mappedBy = "automation", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    @JsonManagedReference
    private List<Keyword> keywords;

    @Column(nullable = false)
    private String userId;
}

File: client/types/automation.ts
import { Listeners, MediaType } from "@/enums/automation";

export type Automation = {
  id: string;
  name: string;
  createdAt: string; // ISO 8601 format date string
  active: boolean;
  triggers: Trigger[];
  listener: Listener | null;
  posts: Post[];
  dms: Dms[];
  keywords: Keyword[];
};

type Trigger = {
  id: string;
  type: string;
};

type Listener = {
  id: string;
  listener: Listeners;
  prompt: string;
  commentReply?: string | null;
  dmCount: number;
  commentCount: number;
};

type Post = {
  id: string;
  postid: string;
  caption?: string;
  media: string;
  mediaType: MediaType;
};

type Dms = {
  id: string;
  senderId?: string;
  receiver?: string;
  message?: string;
};

type Keyword = {
  id: string;
  word: string;
};
